src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+            }
+
+            // disable loading bar
+            this.quickPick.busy = false
+        }
+    }
+}
+
+/**
+ * Represents an iterator that tranforms another iterator into an array of QuickPickItems.
+ * Additionally, has a reset functionality to reset the iterator to its initial state.
+ */
+export class IteratingQuickPickPopulator<TResponse> {
+    private iterator: AsyncIterator<TResponse>
+
+    /**
+     * @param iteratorFactory Function that returns an iterator, with all default state values set. E.g. `collectionUtils.getPaginatedAwsCallIter`
+     * @param transform Function which transforms a response from the iterator into an array of `vscode.QuickPickItem`s.
+     */
+    public constructor(
+        private readonly iteratorFactory: () => AsyncIterator<TResponse>,
+        private readonly transform: (response: TResponse) => vscode.QuickPickItem[]
+    ) {
+        this.iterator = this.iteratorFactory()
+    }
+
+    /**
+     * Resets the iterator to the default state provided by the iteratorFactory.
+     */
+    public reset(): void {
+        this.iterator = this.iteratorFactory()
+    }
+
+    /**
+     * Generates an iterator which returns an array of formatted QuickPickItems on `.next()`
+     */
+    public async *getPickIterator(): AsyncIterator<vscode.QuickPickItem[]> {
+        while (true) {
+            const nextResult = await this.iterator.next()
+            const transformedResult = this.transform(nextResult.value)
+
+            // return (instead of yield) marks final value as done
+            if (nextResult.done) {
+                return transformedResult
+            }
+
+            yield transformedResult
+        }
+    }
+}
+
+/**
+ * Shim function for picker.promptUser calls on pickers that use iteratingQuickPickControllers.
+ * Wraps refresh functionality for this controller and otherwise passes through to a user-provided onDidTriggerButton function
+ * @param button Provided by promptUser
+ * @param resolve Provided by promptUser
+ * @param reject Provided by promptUser
+ * @param iteratingQuickPickController IteratingQuickPickController to call actions on
+ * @param onDidTriggerButton Optional passthrough onDidTriggerButton functionality for promptUser
+ */
+export async function iteratingOnDidTriggerButton<T>(
+    button: vscode.QuickInputButton,
+    resolve: (value: vscode.QuickPickItem[] | PromiseLike<vscode.QuickPickItem[] | undefined> | undefined) => void,
+    reject: (reason?: any) => void,
+    iteratingQuickPickController: IteratingQuickPickController<T>,
+    onDidTriggerButton?: (
+        button: vscode.QuickInputButton,
+        resolve: (value: vscode.QuickPickItem[] | PromiseLike<vscode.QuickPickItem[] | undefined> | undefined) => void,
+        reject: (reason?: any) => void
+    ) => Promise<vscode.QuickPickItem[] | undefined>
+): Promise<vscode.QuickPickItem[] | undefined> {
+    switch (button) {
+        case IteratingQuickPickController.REFRESH_BUTTON:
```

Comment:
Would see if you could move this into the controller somehow. It's awkward that this function knows so much about the members of `IteratingQuickPickController`. It has high cohesion there.

You can also hide the reset operation from the caller.

Reply:
I'd love to as well, but the only way to do that is to create it when the class is created which will pollute that constructor (with the onDidTriggerButton override).

This goes back to how we're currently handling prompts: creating the quick pick, prompting, and handling responses are all separate actions, which I'm not sure is the right pattern but requires a reimagining that is way out of scope for this PR.

Reply:
>  but the only way to do that is to create it when the class is created which will pollute that constructor (with the onDidTriggerButton override).

The issue is similar to what we discussed where the QuickPick is used outside of this class and inside of this class at the same time, leading to unexpected results. If you can hand off all quickpick responsibilities to this class, it will avoid that issue.

Consider two ways to implement this

1. Making the `onDidTriggerButton` a constructor param as you mentioned. This is synonymous with the `showQuickPick()` function which provides `QuickPickOptions`

2. Making a method `onDidTriggerButton()` on this class itself. This is synonymous with [QuickPick#onDidTriggerButton](https://code.visualstudio.com/api/references/vscode-api#QuickPick%3CT%3E)

tl;dr if you have a quickpick item, you should aim to have only *one* actor changing its state to avoid unexpected behavior. Either your class takes control over it entirely (it's a wrapper), or your class provides "suggestions" and the original caller applies those suggestions to the quick pick (your class becomes stateless). This applies pretty generally to stateful objects that would lead to surprising results if used concurrently by two different actors ^

Reply:
2 makes sense if we didn't have our existing promptUser wrapper function. I'm not huge on it but again that would require a larger refactor. 

1 works but it'll make the constructor awfully heavy. Might show you offline what it looks like.